File: frontend/src/types/project.ts
// --- Enums ---

export type RequirementType = 'fr' | 'qa' | 'constraints' | 'other';

export type ProjectModule = 'requirements' | 'design' | 'testcase';

export type ProjectStatus = 'active' | 'archived';

// --- Project ---

export interface ProjectReadiness {
  knowledge: number;
  glossary: number;
  sections: number;
  is_ready: boolean;
}

export interface Project {
  project_id: string;
  name: string;
  description: string | null;
  domain: string | null;
  product_type: string | null;
  modules: ProjectModule[];
  member_count: number;
  status: ProjectStatus;
  readiness: ProjectReadiness | null;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string | null;
  domain?: string | null;
  product_type?: string | null;
  modules: ProjectModule[];
}

export interface ProjectUpdate {
  name?: string | null;
  description?: string | null;
  domain?: string | null;
  product_type?: string | null;
  modules?: ProjectModule[] | null;
}

export interface ProjectListResponse {
  projects: Project[];
}

// --- Requirement ---

// --- Section (요구사항 그룹) ---

export interface Section {
  section_id: string;
  name: string;
  type: string;
  description: string | null;
  output_format_hint: string | null;
  is_default: boolean;
  is_active: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface SectionCreate {
  name: string;
  type: string;
  description?: string | null;
  output_format_hint?: string | null;
}

export interface SectionUpdate {
  name?: string | null;
  description?: string | null;
  output_format_hint?: string | null;
}

export interface SectionReorderRequest {
  ordered_ids: string[];
}

export interface SectionListResponse {
  sections: Section[];
}

export interface Requirement {
  requirement_id: string;
  display_id: string;
  order_index: number;
  type: RequirementType;
  original_text: string;
  refined_text: string | null;
  is_selected: boolean;
  status: string;
  section_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface RequirementReorderRequest {
  ordered_ids: string[];
}

export interface RequirementCreate {
  type: RequirementType;
  original_text: string;
  section_id?: string | null;
}

export interface RequirementUpdate {
  original_text?: string | null;
  refined_text?: string | null;
  is_selected?: boolean | null;
  section_id?: string | null;
}

export interface RequirementListResponse {
  requirements: Requirement[];
}

export interface RequirementSelectionUpdate {
  requirement_ids: string[];
  is_selected: boolean;
}

export interface RequirementSaveResponse {
  version: number;
  saved_count: number;
  saved_at: string;
}

// --- Glossary ---

export interface GlossaryItem {
  glossary_id: string;
  term: string;
  definition: string;
  product_group: string | null;
  synonyms: string[];
  abbreviations: string[];
  section_tags: string[];
  source_document_id: string | null;
  source_document_name: string | null;
  is_auto_extracted: boolean;
  is_approved: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface GlossaryCreate {
  term: string;
  definition: string;
  product_group?: string | null;
  synonyms?: string[];
  abbreviations?: string[];
  section_tags?: string[];
  source_document_id?: string | null;
}

export interface GlossaryUpdate {
  term?: string | null;
  definition?: string | null;
  product_group?: string | null;
  synonyms?: string[] | null;
  abbreviations?: string[] | null;
  section_tags?: string[] | null;
}

export interface GlossaryListResponse {
  glossary: GlossaryItem[];
}

export interface GlossaryGenerateResponse {
  generated_glossary: GlossaryCreate[];
}

export interface GlossaryExtractedItem {
  term: string;
  definition: string;
  synonyms: string[];
  abbreviations: string[];
  source_document_id: string | null;
  source_document_name: string | null;
}

export interface GlossaryExtractResponse {
  candidates: GlossaryExtractedItem[];
}

// --- Review ---

export interface ReviewRequest {
  requirement_ids: string[];
}

export interface ReviewIssue {
  issue_id: string;
  type: 'conflict' | 'duplicate';
  description: string;
  related_requirements: string[];
  hint: string; // 해결 힌트 1줄
}

export interface ReviewSummary {
  total_issues: number;
  conflicts: number;
  duplicates: number;
  ready_for_next: boolean;
  feedback: string;
}

export interface ReviewResponse {
  review_id: string;
  issues: ReviewIssue[];
  summary: ReviewSummary;
}

export interface LatestReviewResponse {
  review_id: string;
  created_at: string;
  reviewed_requirement_ids: string[];
  issues: ReviewIssue[];
  summary: ReviewSummary;
}

// --- Project Settings ---

export interface ProjectSettings {
  llm_model: string;
  language: string;
  export_format: string;
  diagram_tool: string;
}

export interface ProjectSettingsUpdate {
  llm_model?: string | null;
  language?: string | null;
  export_format?: string | null;
  diagram_tool?: string | null;
}

// --- Knowledge Document ---

export type KnowledgeDocumentFileType = 'pdf' | 'md' | 'txt';
export type KnowledgeDocumentStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface KnowledgeDocument {
  document_id: string;
  project_id: string;
  name: string;
  file_type: KnowledgeDocumentFileType;
  size_bytes: number;
  status: KnowledgeDocumentStatus;
  is_active: boolean;
  error_message: string | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeDocumentListResponse {
  documents: KnowledgeDocument[];
  total: number;
}

export interface KnowledgeDocumentPreview {
  document_id: string;
  name: string;
  file_type: KnowledgeDocumentFileType;
  preview_text: string;
  total_characters: number;
}

// --- SRS ---

export interface SrsSection {
  section_id: string | null;
  title: string;
  content: string;
  order_index: number;
}

export interface SrsDocument {
  srs_id: string;
  project_id: string;
  version: number;
  status: string;
  error_message: string | null;
  sections: SrsSection[];
  based_on_records: { artifact_ids?: string[] } | null;
  based_on_documents: { documents?: { id: string; name: string }[] } | null;
  created_at: string;
}

export interface SrsListResponse {
  documents: SrsDocument[];
}

// --- ArtifactRecord (artifact_type='record' 의 도메인 뷰) ---

export type ArtifactRecordStatus = 'draft' | 'approved' | 'excluded';

export interface ArtifactRecord {
  artifact_id: string;
  project_id: string;
  section_id: string | null;
  section_name: string | null;
  content: string;
  display_id: string;
  source_document_id: string | null;
  source_document_name: string | null;
  source_location: string | null;
  confidence_score: number | null;
  status: ArtifactRecordStatus;
  is_auto_extracted: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface ArtifactRecordCreate {
  content: string;
  section_id?: string | null;
  source_document_id?: string | null;
  source_location?: string | null;
}

export interface ArtifactRecordUpdate {
  content?: string | null;
  section_id?: string | null;
}

export interface ArtifactRecordListResponse {
  records: ArtifactRecord[];
  total: number;
}

export interface ArtifactRecordExtractedItem {
  content: string;
  section_id: string | null;
  section_name: string | null;
  source_document_id: string | null;
  source_document_name: string | null;
  source_location: string | null;
  confidence_score: number | null;
}

export interface ArtifactRecordExtractResponse {
  candidates: ArtifactRecordExtractedItem[];
}

// --- Artifact Governance (Git-like: Unstaged/Staged/PR/Merge) ---
export type JsonObject = { [key: string]: unknown };

export type ArtifactKind = 'record' | 'srs' | 'design' | 'testcase';
export type WorkingStatus = 'clean' | 'dirty' | 'staged';
export type LifecycleStatus = 'active' | 'archived' | 'deleted';
export type PullRequestStatus =
  | 'open'
  | 'approved'
  | 'rejected'
  | 'merged'
  | 'superseded';
export type DependencyType = 'derives_from' | 'references' | 'covers';
export type ChangeAction =
  | 'created'
  | 'edited'
  | 'staged'
  | 'pr_opened'
  | 'pr_approved'
  | 'pr_merged'
  | 'pr_rejected'
  | 'reverted';

// Record/SRS/Design/TC 각각의 JSON payload shape은 타입별로 다르다.
// 공용 Artifact 타입은 payload를 제네릭으로 유지.
export interface Artifact<T = JsonObject> {
  artifact_id: string;
  project_id: string;
  artifact_type: ArtifactKind;
  display_id: string;
  title: string | null;
  content: T;
  working_status: WorkingStatus;
  lifecycle_status: LifecycleStatus;
  current_version_id: string | null;
  current_version_number: number | null;
  open_pr_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface ArtifactListResponse {
  artifacts: Artifact[];
  total: number;
}

export interface ArtifactVersion<T = JsonObject> {
  version_id: string;
  artifact_id: string;
  artifact_type: ArtifactKind;
  version_number: number;
  parent_version_id: string | null;
  snapshot: T;
  content_hash: string;
  commit_message: string;
  author_id: string;
  committed_at: string;
  merged_from_pr_id: string | null;
}

export interface ArtifactVersionListResponse {
  versions: ArtifactVersion[];
}

// --- Diff ---

export type DiffFormat = 'unified' | 'deepdiff';

export interface DiffHunk {
  op: 'equal' | 'add' | 'delete';
  text: string;
}

export interface DiffFieldEntry {
  field_path: string;
  kind: 'added' | 'removed' | 'modified' | 'unchanged';
  before?: unknown;
  after?: unknown;
  hunks?: DiffHunk[];
}

export interface DiffResult {
  format: DiffFormat;
  base_version_id: string | null;
  head_version_id: string;
  entries: DiffFieldEntry[];
  unified_text: string | null;
}

// --- PullRequest ---

export interface PullRequest {
  pr_id: string;
  project_id: string;
  artifact_id: string;
  artifact_type: ArtifactKind;
  base_version_id: string | null;
  head_version_id: string;
  status: PullRequestStatus;
  title: string;
  description: string | null;
  author_id: string;
  reviewer_id: string | null;
  created_at: string;
  reviewed_at: string | null;
  merged_at: string | null;
  auto_generated: boolean;
}

export interface PullRequestListResponse {
  pull_requests: PullRequest[];
  total: number;
}

export interface PullRequestCreate {
  title: string;
  description?: string | null;
}

// --- ChangeEvent (audit log) ---

export interface ChangeEvent {
  event_id: string;
  project_id: string;
  artifact_id: string | null;
  pr_id: string | null;
  version_id: string | null;
  action: ChangeAction;
  actor: string;
  diff_summary: JsonObject | null;
  impact_summary: JsonObject | null;
  occurred_at: string;
}

// --- Impact graph ---

export type ImpactReason =
  | 'upstream_version_bumped'
  | 'upstream_status_changed'
  | 'upstream_deleted';

export interface ImpactedArtifactRef {
  artifact_id: string;
  artifact_type: ArtifactKind;
  display_id: string;
  reason: ImpactReason;
  pinned_version_number: number | null;
}

export interface ImpactResponse {
  source_artifact_id: string;
  impacted: ImpactedArtifactRef[];
}

// --- Staging buffers (client-side only) ---
//
// Unstaged/Staged edits는 서버에 도달하지 않은 로컬 버퍼.
// sessionStorage에 persist되어 새로고침을 방어한다.
export interface ArtifactDraft<T = JsonObject> {
  artifact_id: string | null; // null = 신규 (아직 서버 ID 없음)
  local_id: string; // 로컬 고유 식별자 (unstaged 관리용)
  artifact_type: ArtifactKind;
  change_type: 'create' | 'update' | 'delete';
  base_version_id: string | null;
  content: T;
  title?: string | null;
  updated_at: string;
}

// --- Readiness ---

export interface ReadinessItem {
  label: string;
  count: number;
  sufficient: boolean;
}

export interface ReadinessResponse {
  knowledge: ReadinessItem;
  glossary: ReadinessItem;
  sections: ReadinessItem;
  is_ready: boolean;
}

// --- Common ---

export interface ErrorDetail {
  code: string;
  message: string;
  detail?: string | null;
}

export interface ErrorResponse {
  error: ErrorDetail;
}

File: backend/src/schemas/api/project.py
from pydantic import BaseModel, Field, model_validator

from .common import ProjectModule

# 허용되는 모듈 조합 (5가지)
VALID_MODULE_SETS: list[set[str]] = [
    {"requirements", "design", "testcase"},  # All
    {"requirements"},                         # Requirements Only
    {"requirements", "design"},               # Requirements + Design
    {"requirements", "testcase"},             # Requirements + Testcase
    {"testcase"},                              # Testcase Only
]


def _validate_modules(modules: list[ProjectModule]) -> list[ProjectModule]:
    given = {m.value for m in modules}
    if given not in VALID_MODULE_SETS:
        allowed = "All, Requirements Only, Requirements+Design, Requirements+Testcase, Testcase Only"
        raise ValueError(f"허용되지 않는 모듈 조합입니다. 가능한 조합: {allowed}")
    return modules


class ProjectCreate(BaseModel):
    """프로젝트 생성 요청"""
    name: str = Field(description="프로젝트 이름")
    description: str | None = Field(default=None, description="프로젝트 설명")
    domain: str | None = Field(default=None, description="도메인 (예: robotics)")
    product_type: str | None = Field(default=None, description="제품 유형 (예: embedded)")
    modules: list[ProjectModule] = Field(description="사용할 모듈 목록")

    @model_validator(mode="after")
    def check_modules(self):
        _validate_modules(self.modules)
        return self


class ProjectUpdate(BaseModel):
    """프로젝트 수정 요청"""
    name: str | None = Field(default=None, description="프로젝트 이름")
    description: str | None = Field(default=None, description="프로젝트 설명")
    domain: str | None = Field(default=None, description="도메인")
    product_type: str | None = Field(default=None, description="제품 유형")
    modules: list[ProjectModule] | None = Field(default=None, description="모듈 목록")

    @model_validator(mode="after")
    def check_modules(self):
        if self.modules is not None:
            _validate_modules(self.modules)
        return self


class ProjectReadiness(BaseModel):
    """프로젝트 준비도 요약 (목록용)"""
    knowledge: int = Field(default=0, description="활성 지식 문서 수")
    glossary: int = Field(default=0, description="승인 용어 수")
    sections: int = Field(default=0, description="활성 섹션 수")
    is_ready: bool = Field(default=False)


class ProjectResponse(BaseModel):
    """프로젝트 응답"""
    project_id: str = Field(description="프로젝트 ID")
    name: str = Field(description="프로젝트 이름")
    description: str | None = Field(default=None)
    domain: str | None = Field(default=None)
    product_type: str | None = Field(default=None)
    modules: list[ProjectModule] = Field(default_factory=list)
    member_count: int = Field(default=0, description="멤버 수")
    status: str = Field(default="active", description="프로젝트 상태")
    readiness: ProjectReadiness | None = Field(default=None, description="준비도")
    created_at: str = Field(description="생성일시")
    updated_at: str = Field(description="수정일시")


class ProjectListResponse(BaseModel):
    """프로젝트 목록 응답"""
    projects: list[ProjectResponse] = Field(default_factory=list)


class ProjectSettingsResponse(BaseModel):
    """프로젝트 설정 응답"""
    llm_model: str = Field(default="gpt-4", description="LLM 모델")
    language: str = Field(default="ko", description="생성 언어")
    export_format: str = Field(default="pdf", description="기본 Export 형식")
    diagram_tool: str = Field(default="plantuml", description="다이어그램 도구")
    polarion_pat: str | None = Field(default=None, description="Polarion PAT (마스킹)")


class ProjectSettingsUpdate(BaseModel):
    """프로젝트 설정 수정"""
    llm_model: str | None = Field(default=None)
    language: str | None = Field(default=None)
    export_format: str | None = Field(default=None)
    diagram_tool: str | None = Field(default=None)
    polarion_pat: str | None = Field(default=None)


class ProjectDeletePreview(BaseModel):
    """프로젝트 삭제 시 영향받을 데이터 카운트.

    soft delete 단계에서는 데이터가 실제로 사라지지 않지만(휴지통),
    hard delete (30일 후 cron 또는 즉시) 시점에는 모두 영구 삭제.
    """

    project_id: str
    project_name: str
    knowledge_documents: int = 0
    knowledge_files_bytes: int = 0  # MinIO 누적 바이트
    sessions: int = 0
    session_messages: int = 0
    artifacts: int = 0
    artifact_versions: int = 0
    pull_requests: int = 0
    glossary_items: int = 0
    requirement_sections: int = 0


class ProjectDeleteRequest(BaseModel):
    """soft delete 호출 시 옵션. confirm_name 으로 type-to-confirm 검증."""

    confirm_name: str | None = Field(
        default=None,
        description=(
            "운영 환경 안전망: 프로젝트 이름과 일치해야 삭제 진행. "
            "비어 있으면 검증 생략(개발용)."
        ),
    )
